[WIP] Implementation proposal

j1elo opened the pull request:
This is a proposal implementation of sdp-bridge. It is a work-in-progress, not thoroughly tested. The focus is on functionality, leaving architecture considerations aside. Specifically:

* Invalid scenarios are just caught and the process ends with a "BUG" message.
* Code is currently written for the WebRtcTransport, not for generic Transports.
* A static RtpCapabilities object (from `BrowserRtpCapabilities`) is used to represent the RTP caps of remote browsers. This should be worked on in the future, to obtain these caps from the SDP Answer that a browser send in response to an SDP Offer. Read more below.
* Some verbose logs are just commented out, instead of using more adequate technique of filtering by logging level.
* "mediasoup" and "mediasoup-client" are added as dependencies in package.json.

An important limitation we've found while writing this code is that there is a "logic loop" in the mediasoup API, which makes it impossible to create a proper SDP endpoint. This is a summary of the issue:

1. In order to send an SDP Offer, the offerer needs to include its own SSRC for the medias that are going to be sent.
2. In mediasoup, the SSRC is obtained from the `Consumer.rtpParameters.encodings[].ssrc`, so i.e. a Consumer must have been created beforehand.
3. In order to create a Consumer, you need to pass its constructor ([Transport.consume()](https://mediasoup.org/documentation/v3/mediasoup/api/#transport-consume)) an `RtpCapabilities` that describes the capabilities of the remote peer.
4. The `RtpCapabilities` that describe what a browser is capable of, would be obtained from an SDP Answer sent by it.
5. To make a browser send an SDP Answer, you have to first send it an SDP Offer. GOTO 1.

We're breaking this "loop" by hardcoding in the source code an RtpCapabilities that represents a typical SDP Answer from the browser, with 1 audio and 1 video stream. There are capabilities for Firefox, Safari, and Chrome, of which currently the Chrome ones are in use because they seem to be compatible with all other browsers, Firefox included.

Apart from that, another addition to the API was the `addConsumer` method. Just like sdp-bridge is able to generate an array `Producer[]` from an SDP Offer, it should also be able to generate an SDP Offer from an array `Consumer[]`, so it was necessary to add a method that allows providing these Consumers.

All in all, this only intends to be a preliminary implementation that opens up further design decisions and conversation.

j1elo committed:
[WIP] Implement initially proposed API

This is a proposal implementation of sdp-bridge. It is a work-in-
progress, not thoroughly tested. The focus is on functionality, leaving
architecture considerations aside. Specifically:

* Invalid scenarios are just caught and the process ends with a "BUG"
message.
* Code is currently written for the WebRtcTransport, not for generic
Transports.
* A static RtpCapabilities object (from `BrowserRtpCapabilities`) is
used to represent the RTP caps of remote browsers. This should be worked
on in the future, to obtain these caps from the SDP Answer that a
browser send in response to an SDP Offer. Read more below.
* Some verbose logs are just commented out, instead of using more
adequate technique of filtering by logging level.
* "mediasoup" and "mediasoup-client" are added as dependencies in
package.json.

An important limitation we've found while writing this code is that
there is a "logic loop" in the mediasoup API, which makes it impossible
to create a proper SDP endpoint. This is a summary of the issue:

1. In order to send an SDP Offer, the offerer needs to include its own
SSRC for the medias that are going to be sent.
2. In mediasoup, the SSRC is obtained from the
`Consumer.rtpParameters.encodings[].ssrc`, so a Consumer must have been
created beforehand.
3. In order to create a Consumer, you need to pass its constructor an
`RtpCapabilities` that describes the capabilities of the remote peer.
4. The `RtpCapabilities` that describe what a browser is capable of,
would be obtained from an SDP Answer sent by it.
5. To make a browser send an SDP Answer, you have to first send it an
SDP Offer. GOTO 1.

We're breaking this "loop" by hardcoding in the source code an
RtpCapabilities that represents a typical SDP Answer from the browser,
with 1 audio and 1 video stream. There are capabilities for Firefox,
Safari, and Chrome, of which currently the Chrome ones are in use
because they seem to be compatible with all other browsers, Firefox
included.

Apart from that, another addition to the API was the `addConsumer`
method. Just like sdp-bridge is able to generate an array `Producer[]`
from an SDP Offer, it should also be able to generate an SDP Offer from
an array `Consumer[]`, so it was necessary to add a method that allows
providing these Consumers.

All in all, this only intends to be a preliminary implementation that
opens up further design decisions and conversation.

j1elo committed:
Add Git & Prettier config

j1elo committed:
"mediasoup" dependency set only in peerDependencies (tested with 3.6)

j1elo commented:
Please ignore notifications of WIP commits while we play with this

j1elo committed:
[WIP] Fix last bugs

It is now working; verified with openvidu-mediasoup-controller.

j1elo committed:
tsconfig review

* Extend config for Node 10 (official TSConfig Bases).
* Remove settings that were enabled anyway by default.
* Add "removeComments" and "sourceMap".
* Sort alphabetically and format (default settings).
* Review and fix all compilation errors that were caused by
'strictNullChecks = true', caused implicitly by 'strict = true'.

j1elo committed:
tsconfig review

j1elo committed:
SdpUtils: add parsing of "rids" from simulcast SDP

j1elo committed:
Log creation of Producers

j1elo committed:
package.json: add targets build:watch, clean, rebuild, format

j1elo committed:
package.json: move '@types/*' to devDependencies

j1elo committed:
Add lib/ so it's possible to depend on this project by Git URL

piranna reviewed (changes requested):
This change is not needed at all, by using a npm `prepare` script [it will automatically build](https://docs.npmjs.com/cli/v6/using-npm/scripts#prepare-and-prepublish) the `lib/` folder version when using as a git dependency when doing `npm install`.

piranna commented on package.json:
```diff
@@ -15,12 +15,17 @@
-  "main": "lib/index.js",
-  "types": "lib/index.d.ts",
+  "main": "lib/main.js",
+  "types": "lib/main.d.ts",
```
Standard filename is `index.js`, revert this.

j1elo replied:
Done

piranna commented on src/main.ts:
```diff
@@ -0,0 +1,297 @@
+        "[SdpEndpoint.processOffer] ERROR: A remote description was already set"
+      );
+      return [];
+    }
```
RTCPeerConnection state machine allows to process an offer after having a remote SDP depending on what state we are, probably we should implement that too.

j1elo committed:
package.json: update to mediasoup 3.7

Also sort dependencies alphabetically

j1elo committed:
delete lib/ and use a prepare script instead

If a package being installed through git contains a prepare script, its
dependencies and devDependencies will be installed, and the prepare
script will be run, before the package is packaged and installed.

https://docs.npmjs.com/cli/v7/using-npm/scripts#life-cycle-scripts

j1elo committed:
rename 'main.ts' to more standard 'index.ts'

j1elo committed:
add tool "browser-rtpcapabilities"

This tool makes it trivially easy to obtain the RtpCapabilities object
that represents what encodings and other WebRTC features are available
on the web browser the tool runs on.

j1elo committed:
update RtpCapabilities from Firefox 88, Chrome 91, Safari 14.1

j1elo committed:
README: full review, add section Implementation Notes

j1elo committed:
README: small additions

j1elo committed:
Bump to Typescript 4 and mediasoup 3.8

j1elo commented:
Closing this until the fork is in a good shape for contribution:

https://github.com/OpenVidu/mediasoup-sdp-bridge
